### ISSUE TITLE
AJP throws NPE for responses larger than 8192 bytes

### ISSUE THREAD

**Issue by glassfishrobot** [eclipse-ee4j/glassfish#22295]
## Environment Details
* GlassFish Server Open Source Edition  5.0  (build 25)
* JDK version: 1.8.0_144
* OS: Linux

----------

## Problem Description
When using AJP between an Apache webserver and GlassFish, GlassFish throws a NullPointerException after writing the response chunks for a request, instead of sending an AJP END_RESPONSE packet. This only happens when the response body is larger than 8192 bytes.

Increasing the send buffer size of the AJP listener (which defaults to 8192 bytes) doesn't make any difference.

```
java.lang.NullPointerException
	at org.glassfish.grizzly.http.ajp.AjpHandlerFilter.encodeHttpPacket(AjpHandlerFilter.java:282)
	at org.glassfish.grizzly.http.ajp.AjpHandlerFilter.handleWrite(AjpHandlerFilter.java:245)
	at org.glassfish.grizzly.filterchain.ExecutorResolver$8.execute(ExecutorResolver.java:111)
	at org.glassfish.grizzly.filterchain.DefaultFilterChain.executeFilter(DefaultFilterChain.java:284)
	at org.glassfish.grizzly.filterchain.DefaultFilterChain.executeChainPart(DefaultFilterChain.java:201)
	at org.glassfish.grizzly.filterchain.DefaultFilterChain.execute(DefaultFilterChain.java:133)
	at org.glassfish.grizzly.filterchain.DefaultFilterChain.process(DefaultFilterChain.java:112)
	at org.glassfish.grizzly.ProcessorExecutor.execute(ProcessorExecutor.java:77)
	at org.glassfish.grizzly.filterchain.FilterChainContext.write(FilterChainContext.java:890)
	at org.glassfish.grizzly.filterchain.FilterChainContext.write(FilterChainContext.java:858)
	at org.glassfish.grizzly.http.io.OutputBuffer.flushBuffer(OutputBuffer.java:1092)
	at org.glassfish.grizzly.http.io.OutputBuffer.flushBinaryBuffers(OutputBuffer.java:1066)
	at org.glassfish.grizzly.http.io.OutputBuffer.flushAllBuffers(OutputBuffer.java:1036)
	at org.glassfish.grizzly.http.io.OutputBuffer.close(OutputBuffer.java:754)
	at org.glassfish.grizzly.http.io.OutputBuffer.endRequest(OutputBuffer.java:401)
	at org.glassfish.grizzly.http.server.Response.finish(Response.java:559)
	at org.glassfish.grizzly.http.server.HttpServerFilter.afterService(HttpServerFilter.java:397)
	at org.glassfish.grizzly.http.server.HttpServerFilter.handleRead(HttpServerFilter.java:267)
	at org.glassfish.grizzly.filterchain.ExecutorResolver$9.execute(ExecutorResolver.java:119)
	at org.glassfish.grizzly.filterchain.DefaultFilterChain.executeFilter(DefaultFilterChain.java:284)
	at org.glassfish.grizzly.filterchain.DefaultFilterChain.executeChainPart(DefaultFilterChain.java:201)
	at org.glassfish.grizzly.filterchain.DefaultFilterChain.execute(DefaultFilterChain.java:133)
	at org.glassfish.grizzly.filterchain.DefaultFilterChain.process(DefaultFilterChain.java:112)
	at org.glassfish.grizzly.ProcessorExecutor.execute(ProcessorExecutor.java:77)
	at org.glassfish.grizzly.nio.transport.TCPNIOTransport.fireIOEvent(TCPNIOTransport.java:539)
	at org.glassfish.grizzly.strategies.AbstractIOStrategy.fireIOEvent(AbstractIOStrategy.java:112)
	at org.glassfish.grizzly.strategies.WorkerThreadIOStrategy.run0(WorkerThreadIOStrategy.java:117)
	at org.glassfish.grizzly.strategies.WorkerThreadIOStrategy.access$100(WorkerThreadIOStrategy.java:56)
	at org.glassfish.grizzly.strategies.WorkerThreadIOStrategy$WorkerThreadRunnable.run(WorkerThreadIOStrategy.java:137)
	at org.glassfish.grizzly.threadpool.AbstractThreadPool$Worker.doWork(AbstractThreadPool.java:593)
	at org.glassfish.grizzly.threadpool.AbstractThreadPool$Worker.run(AbstractThreadPool.java:573)
	at java.lang.Thread.run(Thread.java:748)
```

## Steps to reproduce
* Configure an AJP listener in GlassFish
* Configure Apache httpd to proxy requests to GlassFish using mod_proxy_ajp
* Request an URL with a response body larger than 8192 bytes

## Impact of Issue
AJP cannot be used anymore.

**Comment by javierllorente** [comment]
Unfortunately, this issue affects GlassFish 5.1.0 as well; mod_jk/mod_proxy_ajp doesn't work properly unless you manually patch Grizzly. Detailed instructions: https://stackoverflow.com/a/49108723

I think it should be mentioned in "Known Issues in GlassFish Server 5.1".
https://glassfish.org/docs/latest/release-notes/release-notes.html

**Comment by dmatej** [comment]
I believe it was fixed later, so in GlassFish 6 it should not happen any more, can you recheck it for the latest GlassFish version, please?

**Comment by javierllorente** [comment]
@dmatej Same issue on GlassFish 6.0.0
```
[2021-07-12T13:50:15.059+0200] [glassfish 6.0] [WARNING] [] [org.glassfish.grizzly.filterchain.DefaultFilterChain] [tid: _ThreadID=56 _ThreadName=glassfish-listener(3)] [timeMillis: 1626090615059] [levelValue: 900] [[

  GRIZZLY0013: Exception during FilterChain execution
java.lang.NullPointerException
        at org.glassfish.grizzly.http.ajp.AjpHandlerFilter.encodeHttpPacket(AjpHandlerFilter.java:232)
        at org.glassfish.grizzly.http.ajp.AjpHandlerFilter.handleWrite(AjpHandlerFilter.java:198)
        at org.glassfish.grizzly.filterchain.ExecutorResolver$8.execute(ExecutorResolver.java:81)
        at org.glassfish.grizzly.filterchain.DefaultFilterChain.executeFilter(DefaultFilterChain.java:246)
        at org.glassfish.grizzly.filterchain.DefaultFilterChain.executeChainPart(DefaultFilterChain.java:178)
        at org.glassfish.grizzly.filterchain.DefaultFilterChain.execute(DefaultFilterChain.java:118)
        at org.glassfish.grizzly.filterchain.DefaultFilterChain.process(DefaultFilterChain.java:96)
        at org.glassfish.grizzly.ProcessorExecutor.execute(ProcessorExecutor.java:51)
        at org.glassfish.grizzly.filterchain.FilterChainContext.write(FilterChainContext.java:750)
        at org.glassfish.grizzly.filterchain.FilterChainContext.write(FilterChainContext.java:728)
        at org.glassfish.grizzly.http.io.OutputBuffer.flushBuffer(OutputBuffer.java:1004)
        at org.glassfish.grizzly.http.io.OutputBuffer.flushBinaryBuffers(OutputBuffer.java:980)
        at org.glassfish.grizzly.http.io.OutputBuffer.flushAllBuffers(OutputBuffer.java:952)
        at org.glassfish.grizzly.http.io.OutputBuffer.close(OutputBuffer.java:677)
        at org.apache.catalina.connector.OutputBuffer.close(OutputBuffer.java:204)
        at org.apache.catalina.connector.CoyoteOutputStream.close(CoyoteOutputStream.java:138)
        at java.nio.channels.Channels$WritableByteChannelImpl.implCloseChannel(Channels.java:469)
        at java.nio.channels.spi.AbstractInterruptibleChannel.close(AbstractInterruptibleChannel.java:115)
        at com.sun.faces.application.resource.ResourceHandlerImpl.handleResourceRequest(ResourceHandlerImpl.java:325)
        at jakarta.faces.application.ResourceHandlerWrapper.handleResourceRequest(ResourceHandlerWrapper.java:166)
        at org.primefaces.application.resource.PrimeResourceHandler.handleResourceRequest(PrimeResourceHandler.java:87)
        at jakarta.faces.application.ResourceHandlerWrapper.handleResourceRequest(ResourceHandlerWrapper.java:166)
        at jakarta.faces.webapp.FacesServlet.executeLifecyle(FacesServlet.java:678)
        at jakarta.faces.webapp.FacesServlet.service(FacesServlet.java:437)
        at org.apache.catalina.core.StandardWrapper.service(StandardWrapper.java:1540)
        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:297)
        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:167)
        at org.primefaces.webapp.filter.FileUploadFilter.doFilter(FileUploadFilter.java:89)
        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:209)
        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:167)
        at org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:215)
        at org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:119)
        at org.apache.catalina.core.StandardPipeline.doInvoke(StandardPipeline.java:611)
        at org.apache.catalina.core.StandardPipeline.invoke(StandardPipeline.java:550)
        at com.sun.enterprise.web.WebPipeline.invoke(WebPipeline.java:75)
        at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:114)
        at org.apache.catalina.connector.CoyoteAdapter.doService(CoyoteAdapter.java:332)
        at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:199)
        at com.sun.enterprise.v3.services.impl.ContainerMapper$HttpHandlerCallable.call(ContainerMapper.java:440)
        at com.sun.enterprise.v3.services.impl.ContainerMapper.service(ContainerMapper.java:144)
        at org.glassfish.grizzly.http.server.HttpHandler.runService(HttpHandler.java:174)
        at org.glassfish.grizzly.http.server.HttpHandler.doHandle(HttpHandler.java:153)
        at org.glassfish.grizzly.http.server.HttpServerFilter.handleRead(HttpServerFilter.java:196)
        at org.glassfish.grizzly.filterchain.ExecutorResolver$9.execute(ExecutorResolver.java:88)
        at org.glassfish.grizzly.filterchain.DefaultFilterChain.executeFilter(DefaultFilterChain.java:246)
        at org.glassfish.grizzly.filterchain.DefaultFilterChain.executeChainPart(DefaultFilterChain.java:178)
        at org.glassfish.grizzly.filterchain.DefaultFilterChain.execute(DefaultFilterChain.java:118)
        at org.glassfish.grizzly.filterchain.DefaultFilterChain.process(DefaultFilterChain.java:96)
        at org.glassfish.grizzly.ProcessorExecutor.execute(ProcessorExecutor.java:51)
        at org.glassfish.grizzly.nio.transport.TCPNIOTransport.fireIOEvent(TCPNIOTransport.java:510)
        at org.glassfish.grizzly.strategies.AbstractIOStrategy.fireIOEvent(AbstractIOStrategy.java:82)
        at org.glassfish.grizzly.strategies.WorkerThreadIOStrategy.run0(WorkerThreadIOStrategy.java:83)
        at org.glassfish.grizzly.strategies.WorkerThreadIOStrategy.access$100(WorkerThreadIOStrategy.java:34)
        at org.glassfish.grizzly.strategies.WorkerThreadIOStrategy$WorkerThreadRunnable.run(WorkerThreadIOStrategy.java:101)
        at org.glassfish.grizzly.threadpool.AbstractThreadPool$Worker.doWork(AbstractThreadPool.java:535)
        at org.glassfish.grizzly.threadpool.AbstractThreadPool$Worker.run(AbstractThreadPool.java:515)
        at java.lang.Thread.run(Thread.java:748)```

**Comment by javierllorente** [comment]
I have also tested GlassFish 6.1.0 and 6.2.0 with the same result: ``GRIZZLY0013: Exception during FilterChain execution  java.lang.NullPointerException``

**Comment by javierllorente** [comment]
Manually patching Grizzly makes ajp on GlassFish 6.2.0 work again.

**Comment by javierllorente** [comment]
I can't reproduce this issue on Payara 5.2021.4

**Comment by javierllorente** [comment]
Payara fix: https://github.com/payara/Payara/issues/2649

**Comment by javierllorente** [comment]
@dmatej What do you think of https://github.com/eclipse-ee4j/grizzly/pull/2079? 
It's the same fix as https://github.com/payara/patched-src-grizzly/pull/9. We could ask @omolenkamp to rebase his fix to merge it.

**Comment by dmatej** [comment]
Aha, the fixed grizzly wasn't released yet, that is why this issue is not closed yet!
Fixed in https://github.com/eclipse-ee4j/grizzly/pull/2132

**Comment by dmatej** [comment]
Should be fixed with grizzly 3.0.1 - that was upgraded in https://github.com/eclipse-ee4j/glassfish/pull/23631
